doc/users/next_whats_new/constrained_layout.rst

```diff
@@ -0,0 +1,59 @@
+Constrained Layout Manager
+---------------------------
+
+.. warning::
+
+    Constrained Layout is **experimental**.  The
+    behaviour and API are subject to change, or the whole functionality
+    may be removed without a deprecation period.
+
+
+A new method to automatically decide spacing between subplots and their
+organizing ``GridSpec`` instances has been added.  It is meant to
+replace the venerable ``tight_layout`` method.  It is invoked via
+a new ``plt.figure(constrained_layout=True)`` kwarg to
+`~.figure.Figure` or `~.figure.subplots`.
+
+There are new ``rcParams`` for this package, and spacing can be
+more finely tuned with the new `~.set_constrained_layout_pads`.
+
+Features include:
+
+  - Automatic spacing for subplots with a fixed-size padding in inches around
+    subplots and all their decorators, and space between as a fraction
+    of subplot size between subplots.
+  - Spacing for `~.figure.suptitle`, and colorbars that are attached to
+    more than one axes.
+  - Nested `~.GridSpec` layouts using `~.GridSpecFromSubplotSpec`.
+
+  For more details and capabilities please see the new tutorial:
+  :doc:`/tutorials/intermediate/constrainedlayout_guide`
+
+Note the new API to access this:
+
+New ``plt.figure`` and ``plt.subplots`` kwarg: ``constrained_layout``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:meth:`~matplotlib.pyplot.figure` and :meth:`~matplotlib.pyplot.subplots`
+can now be called with ``constrained_layout=True`` kwarg to enable
+constrained_layout.
+
+New ``ax.set_position`` behaviour
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:meth:`~matplotlib.axes.set_position` now makes the specified axis no
+longer responsive to ``constrained_layout``, consistent with the idea that the
+user wants to place an axis manually.
+
+Internally, this means that old ``ax.set_position`` calls *inside* the library
+are changed to private ``ax._set_position`` calls so that
+``constrained_layout`` will still work with these axes.
+
+New ``fig`` kwarg for ``GridSpec``
```

Comment:
figure=?  (see comment)

Reply:
Oops, sorry, forgot!  I’ll do later tonight...

Reply:
Hmmm.   `gridspec.py` already uses `fig=`.  

https://github.com/matplotlib/matplotlib/blob/b511bb250ffaa7419ee9bc17216ab9bab3f06e5c/lib/matplotlib/gridspec.py#L223

Its a bit of a toss-up in the code about which gets used.  Which I agree is annoying.  Guidance?

Reply:
So I guess 

1. deprecate `fig=` in `gridspec.py` and add `figure=` to `get_subplot_params`.  Honestly I think that's the only place its used, and I don't think thats a heavily used piece of user code.
2. use `fig=` in the new code.

Reply:
Prefer 1).

Reply:
Done, please let me know if I did it properly:

```python
if fig is not None:
            warnings.warn("the 'fig' kwarg is deprecated "
                          "use 'figure' instead", mplDeprecation)
        if figure is None:
            figure = fig
```